File: src/main/scala/Race/common/LineSegment.scala
package Race.common

object LineSegment {

  def apply(p1: Point, p2: Point): LineSegment = LineSegmentImpl(p1, p2)


  def intersect(s: LineSegment, l2: LineSlopeIntercept): Option[Point] = {
    val l1 = s match {
      case local: LineSegmentImpl => local.lineSlope
      case _ => LineSlopeIntercept.tryMake(s.p1, s.p2)
    }

    if (l1.isEmpty) return None

    val intersect: Option[Point] = l1.get.intersect(l2)

    intersect match {
      case Some(p) if s.inSquare(p) => intersect
      case _ => None
    }
  }

  def intersect(s1: LineSegment, s2: LineSegment): Option[Point] = {
    val l1 = s1 match {
      case local: LineSegmentImpl => local.lineSlope
      case _ => LineSlopeIntercept.tryMake(s1.p1, s1.p2)
    }

    val l2 = s2 match {
      case local: LineSegmentImpl => local.lineSlope
      case _ => LineSlopeIntercept.tryMake(s2.p1, s2.p2)
    }

    if (l1.isEmpty || l2.isEmpty) return None

    val intersect: Option[Point] = l1.get.intersect(l2.get)

    intersect match {
      case Some(p) if s1.inSquare(p) && s2.inSquare(p) => intersect
      case _ => None
    }
  }

  def closestPoint(line: LineSegment, point: Point): Option[Point] = {
    if (line.lineStandard.isEmpty) return None
    val p = LineStandardForm.closestPoint(line.lineStandard.get, point)
    Some(p).filter(line.inSquare)
  }

  case class LineSegmentImpl(p1: Point, p2: Point) extends LineSegment {
    lazy val lineSlope: Option[LineSlopeIntercept] = LineSlopeIntercept.tryMake(p1, p2)
    lazy val lineStandard: Option[LineStandardForm] = Some(LineStandardForm(p1, p2))

    override def length: Double = Point.distance(p1, p2)
    override def intersect(other: LineSlopeIntercept): Option[Point] = LineSegment.intersect(this, other)
    override def intersect(other: LineSegment): Option[Point] = LineSegment.intersect(this, other)
    override def inSquare(p: Point): Boolean = Point.inSquare(p, p1, p2)
    override def closestPoint(point: Point): Option[Point] = LineSegment.closestPoint(this, point)
    override def angle: Double = {
      if (lineSlope.isEmpty) return 0
      val angle = lineSlope.get.angle
      if (p2.x > p1.x) angle
      else if (angle > 0) (2 * Math.PI) - angle
      else (-2 * Math.PI) - angle
    }
  }
}

trait LineSegment {
  def p1: Point
  def p2: Point

  def length: Double
  def lineSlope: Option[LineSlopeIntercept]
  def lineStandard: Option[LineStandardForm]
  def intersect(other: LineSlopeIntercept): Option[Point]
  def intersect(other: LineSegment): Option[Point]
  def inSquare(p: Point): Boolean
  def closestPoint(point: Point): Option[Point]
  def angle: Double // from x-axis
}

File: PyCar/common/line_segment.py
from common.line_slope_intercept import LineSlopeIntercept
from common.line_standard_form import LineStandardForm
from common.point import Point
from math import pi


class LineSegment:
    def __init__(self, _p1: Point, _p2: Point):
        self.p1 = _p1
        self.p2 = _p2
        self._line_slope = None
        self._line_standard = None

    def length(self) -> float:
        return Point.distance(self.p1, self.p2)

    def length_from_start(self, p: Point) -> float:
        x = self.closest_point(p)
        return Point.distance(self.p1, x) if x is not None else None

    def length_from_end(self, p: Point) -> float:
        x = self.closest_point(p)
        return Point.distance(self.p2, x) if x is not None else None

    @property
    def line_slope(self) -> LineSlopeIntercept:
        if self._line_slope is None:
            self._line_slope = LineSlopeIntercept.try_make(self.p1, self.p2)
        return self._line_slope

    @property
    def line_standard(self) -> LineStandardForm:
        if self._line_standard is None:
            self._line_standard = LineStandardForm.try_make(self.p1, self.p2)
        return self._line_standard

    def in_box(self, p: Point) -> bool:
        return Point.in_box(p, self.p1, self.p2)

    def intersect(self, other):
        l1 = self.line_slope
        if isinstance(other, LineSlopeIntercept):
            l2 = other
            def check_l2(x): return True
        else:
            l2 = other.line_slope
            def check_l2(x): return other.in_box(x)

        if l1 is None or l2 is None:
            return None
        intersect = l1.intersect(l2)
        if intersect is not None and self.in_box(intersect) and check_l2(intersect):
            return intersect
        else:
            return None

    def closest_point(self, p: Point):
        lsf = self.line_standard
        if lsf is None:
            return None
        cp = lsf.closest_point(p)
        if cp is not None and self.in_box(cp):
            return cp
        return None

    def angle(self) -> float:
        ls = self.line_slope
        if ls is None:
            return 0
        angle = ls.angle()
        if self.p2.x() > self.p1.x(): return angle
        else:
            if angle > 0:
                return (2 * pi) - angle
            else:
                return (-2 * pi) - angle

